Fix validation for Haplotypes

MillironX opened the pull request:
## Types of changes

This PR implements the following changes:
_(Please tick any or all of the following that are applicable)_

- [ ] :sparkles: New feature (A non-breaking change which adds functionality).
- [x] :bug: Bug fix (A non-breaking change, which fixes an issue).
- [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail

Fixes #36 

I've added a ton of validation tests and refactored the `_is_valid(::Haplotype)` function to provide correctness.

### Runnable example

```julia
using BioSequences, SequenceVariation, BenchmarkTools
seq2 = dna"TGATGCGTGTAGCAACACTTATAGCG"

invalid_haplotype = Haplotype(
    seq2, [Variation(seq2, "T2A"), Variation(seq2, "T2C")]
)

@show invalid_haplotype

valid_haplotype = Haplotype(
    seq2,
    [
        Variation(seq2, "0AAA"),
        Variation(seq2, "T2A"),
        Variation(seq2, "Δ5-8"),
        Variation(seq2, "G26A"),
    ],
)

@benchmark SequenceVariation._is_valid(valid_haplotype)
```

#### Previous Results

```plaintext
invalid_haplotype = Haplotype{LongSequence{DNAAlphabet{4}}, DNA} with 2 edits:
  G2A
  G2C
BenchmarkTools.Trial: 10000 samples with 997 evaluations.
 Range (min … max):  22.028 ns …  2.094 μs  ┊ GC (min … max): 0.00% … 94.65%
 Time  (median):     23.143 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   25.932 ns ± 51.773 ns  ┊ GC (mean ± σ):  5.14% ±  2.55%

  ▄█▆▆▆▅▅▄▄▄▄▃▂            ▁ ▁▂▂▂▂▂▂▁                         ▂
  ██████████████▇▇▇▇▇▅▆▇▇▇▇█████████████▇█▇███▇▆▆▆▆▆▆▆▆▆▆▆▆▅▅ █
  22 ns        Histogram: log(frequency) by time      37.4 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```

#### This PR Results

```plaintext
ERROR: LoadError: Multiple modifications at same position
[...]
BenchmarkTools.Trial: 10000 samples with 995 evaluations.
 Range (min … max):  28.305 ns …  2.353 μs  ┊ GC (min … max): 0.00% … 97.24%
 Time  (median):     29.393 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   32.322 ns ± 59.284 ns  ┊ GC (mean ± σ):  4.75% ±  2.56%

   ▃▆█▇▆▄▃▁▁                    ▁▁▂▁▂▂▁▁                      ▂
  ▅███████████▇▇▇▇█▆▅▅▆▆▇▇▇█████████████████▆█▆▆▆▆▅▅▅▅▆▆▇▆▆▅▆ █
  28.3 ns      Histogram: log(frequency) by time      43.7 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```

## :ballot_box_with_check: Checklist

- [x] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/).
- [x] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/v1/manual/documentation/).
- [x] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [x] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [x] :ok: All changes should be compatible with the latest stable version of Julia.
- [x] :thought_balloon: I have commented liberally for any complex pieces of internal code.

kescobo reviewed (changes requested):
LGTM - failing test on julia 1.6 seems to be coming [from here](https://github.com/BioJulia/SequenceVariation.jl/pull/37/files#diff-e84e296f931a7cae11eb0b714d8285a9b6897ef391c0a01732c09e6665f7e635R69)

Not sure what has changed about `UnitRange()` that would affect this, but it seems pretty fundamental. Can this be refactored to something that will work on 1.6? I think it should be as easy as explicitly converting `length(reference(h))` to UInt64:

```
julia> VERSION
v"1.6.7"

julia> x = [1,2,3];

julia> UnitRange(UInt64(1), length(x))
ERROR: MethodError: no method matching UnitRange(::UInt64, ::Int64)
Closest candidates are:
  UnitRange(::T, ::T) where T<:Real at range.jl:289
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1

julia> UnitRange(UInt64(1), UInt64(length(x)))
0x0000000000000001:0x0000000000000003
```

```
julia> VERSION
v"1.8.5"

julia> x = [1,2,3];

julia> UnitRange(UInt64(1), length(x))
0x0000000000000001:0x0000000000000003
```

MillironX commented:
> failing test on julia 1.6 seems to be coming [from here](https://github.com/BioJulia/SequenceVariation.jl/pull/37/files#diff-e84e296f931a7cae11eb0b714d8285a9b6897ef391c0a01732c09e6665f7e635R69)

No kidding. Unfortunately I've been kept away from my Julia dev machine since I submitted

> I think it should be as easy as explicitly converting `length(reference(h))` to UInt64

I'm glad it's that simple: I was expected something much more involved. I will test this myself tomorrow morning.

MillironX commented on src/Haplotype.jl:
```diff
@@ -49,41 +49,60 @@ function Base.show(io::IO, x::Haplotype)
+    # is because we couldn't tell if a starting insertion overlapped with a future edit that
+    # way. Also, force the range to be a signed integer to avoid subtraction overflow errors
+    # when there is a starting insertion.
+    valid_positions = UnitRange{Int128}(leftposition(first(_edits(h))), length(reference(h)))
```
```suggestion
    valid_positions = UnitRange{Int128}(
        leftposition(first(_edits(h))), UInt64(length(reference(h)))
    )
```

MillironX commented on src/Haplotype.jl:
```diff
@@ -49,41 +49,60 @@ function Base.show(io::IO, x::Haplotype)
         # Sanity check: for this to be a valid variant, it must be comprised of valid
         # variations
-        _is_valid(Variation(h.ref, edit)) || return (false, "Invalid Variation")
+        _is_valid(Variation{S,T}(h.ref, edit, Unsafe())) || return (false, "Invalid Variation")
```
```suggestion
        _is_valid(Variation{S,T}(h.ref, edit, Unsafe())) || 
            return (false, "Invalid Variation")
```

MillironX commented on src/Haplotype.jl:
```diff
@@ -49,41 +49,60 @@ function Base.show(io::IO, x::Haplotype)
+function _is_valid(h::Haplotype{S,T}) where {S,T}
+    # Empty references mean we have nothing to compare to
+    isempty(reference(h)) && return (false, "Empty reference")
+    
```
```suggestion

```

MillironX commented on src/Haplotype.jl:
```diff
@@ -49,41 +49,62 @@ function Base.show(io::IO, x::Haplotype)
+    valid_positions = UnitRange{Int128}(
+        leftposition(first(_edits(h))), UInt64(length(reference(h)))
+    )
+    
```
```suggestion

```

MillironX commented on src/Haplotype.jl:
```diff
@@ -49,41 +49,62 @@ function Base.show(io::IO, x::Haplotype)
         # Sanity check: for this to be a valid variant, it must be comprised of valid
         # variations
-        _is_valid(Variation(h.ref, edit)) || return (false, "Invalid Variation")
+        _is_valid(Variation{S,T}(h.ref, edit, Unsafe())) || 
```
```suggestion
        _is_valid(Variation{S,T}(h.ref, edit, Unsafe())) ||
```

kescobo reviewed (approved):
:+1: Would be nice to not have an :x:  on codecov, but I don't consider that blocking. Thanks for tackling this!

MillironX commented:
> Would be nice to not have an x on codecov, but I don't consider that blocking

The irony is that I've added direct tests where there was only indirect coverage before. I think it's "failing" because I've added more lines of code to test, even if every added line is tested.

I'm going to squash some of these commits, add a CHANGELOG entry and merge.
